File: core/src/test/java/meshu/core/l3/CborTest.java
package meshu.core.l3;

import org.junit.jupiter.api.Test;

import java.util.HexFormat;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

/**
 * TESTVECTORS.md vectors 4, 8 (§5, §8.2, §10): CBOR envelope + canonical form.
 */
class CborTest {

    private static final HexFormat HEX = HexFormat.of();

    /** Vector 4: MINT_QUOTE request must encode to exactly 42 bytes. */
    @Test
    void mintQuoteRequestEncodesToVector4() {
        // [ 0x01, 1, 1000, 0, h'02a1a1…a1' (33 bytes) ]
        byte[] quotePubkey = new byte[33];
        quotePubkey[0] = 0x02;
        java.util.Arrays.fill(quotePubkey, 1, 33, (byte) 0xa1);

        Envelope env = Envelope.of(Op.MINT_QUOTE,
                Cbor.uint(1),            // mint_handle
                Cbor.uint(1000),         // amount
                Cbor.uint(0),            // unit sat
                Cbor.bytes(quotePubkey)); // quote_pubkey

        byte[] encoded = env.encode();
        String expected = "8501011903e800582102" + "a1".repeat(32);
        assertEquals(expected, HEX.formatHex(encoded));
        assertEquals(42, encoded.length);
    }

    /** Vector 8: ERROR response must encode to exactly 7 bytes. */
    @Test
    void errorResponseEncodesToVector8() {
        // [0xFE, 20001, 0]
        Envelope env = Envelope.of(Op.ERROR, Cbor.uint(20001), Cbor.uint(0));
        byte[] encoded = env.encode();
        assertEquals("8318fe194e2100", HEX.formatHex(encoded));
        assertEquals(7, encoded.length);
    }

    /** Envelope decode recovers opcode as first array element. */
    @Test
    void envelopeRoundTrip() {
        Envelope env = Envelope.of(Op.HELLO,
                Cbor.uint(1), Cbor.bytes(new byte[32]), Cbor.array(Cbor.uint(0x0D)));
        Envelope decoded = Envelope.decode(env.encode());
        assertEquals(Op.HELLO, decoded.opcode());
        assertEquals(Op.HELLO, decoded.opcodeHint());
        assertEquals(3, decoded.fields().size());
    }

    /** §5.2: decoders ignore unknown trailing array elements. */
    @Test
    void ignoresUnknownTrailingElements() {
        // A v2 message with an extra field must still decode as a v1 envelope.
        byte[] v2 = Cbor.encode(Cbor.array(
                Cbor.uint(Op.KEYSETS), Cbor.uint(0), Cbor.uint(999) /* future field */));
        Envelope env = Envelope.decode(v2);
        assertEquals(Op.KEYSETS, env.opcode());
        assertEquals(2, env.fields().size()); // caller reads what it knows
    }

    /** §5.2: indefinite-length items MUST be rejected. */
    @Test
    void rejectsIndefiniteLength() {
        // 0x9F = indefinite-length array start
        assertThrows(IllegalArgumentException.class,
                () -> Cbor.decode(new byte[]{(byte) 0x9F, 0x01, (byte) 0xFF}));
    }

    /** §5.1: maps are forbidden. */
    @Test
    void rejectsMaps() {
        // 0xA1 = map(1)
        assertThrows(IllegalArgumentException.class,
                () -> Cbor.decode(new byte[]{(byte) 0xA1, 0x01, 0x02}));
    }

    /** §5.2: negative integers unsupported. */
    @Test
    void rejectsNegativeInt() {
        // 0x20 = negative int -1
        assertThrows(IllegalArgumentException.class, () -> Cbor.decode(new byte[]{0x20}));
    }

    @Test
    void canonicalMinimalIntEncoding() {
        // 1000 encodes as 19 03e8 (2-byte), not 1a 000003e8 (4-byte).
        assertArrayEquals(new byte[]{0x19, 0x03, (byte) 0xe8}, Cbor.encode(Cbor.uint(1000)));
        assertArrayEquals(new byte[]{0x17}, Cbor.encode(Cbor.uint(23)));
        assertArrayEquals(new byte[]{0x18, 0x18}, Cbor.encode(Cbor.uint(24)));
        assertArrayEquals(new byte[]{0x00}, Cbor.encode(Cbor.uint(0)));
    }

    @Test
    void decodeRejectsTrailingBytes() {
        assertThrows(IllegalArgumentException.class,
                () -> Cbor.decode(new byte[]{0x01, 0x02}));
    }

    // ---- adversarial: attacker-controlled lengths must not allocate (review C1)

    @Test
    void rejectsOversizedByteStringLength() {
        // 0x5A = byte string with 4-byte length claiming 0x10000000; input is tiny.
        byte[] frame = new byte[]{(byte) 0x5A, 0x10, 0x00, 0x00, 0x00, 0x41};
        assertThrows(IllegalArgumentException.class, () -> Cbor.decode(frame));
    }

    @Test
    void rejectsOversizedTextLength() {
        // 0x7A = text string with 4-byte length claiming 2 GB.
        byte[] frame = new byte[]{(byte) 0x7A, 0x7F, (byte) 0xFF, (byte) 0xFF, (byte) 0xFF, 0x61, 0x62};
        assertThrows(IllegalArgumentException.class, () -> Cbor.decode(frame));
    }

    @Test
    void rejectsHugeArrayCount() {
        // Array count claims ~2 billion items; input has 1 byte.
        byte[] frame = new byte[]{(byte) 0x9A, 0x7F, (byte) 0xFF, (byte) 0xFF, (byte) 0xFF, 0x01};
        assertThrows(IllegalArgumentException.class, () -> Cbor.decode(frame));
    }

    @Test
    void rejectsTruncatedLengthHeader() {
        // Claims a 4-byte length argument but only 1 byte is present.
        byte[] frame = new byte[]{(byte) 0x5A, 0x00};
        assertThrows(IllegalArgumentException.class, () -> Cbor.decode(frame));
    }

    @Test
    void byteStringAndTextRoundTrip() {
        byte[] data = {0x01, 0x02, 0x03};
        Cbor.Value v = Cbor.decode(Cbor.encode(Cbor.bytes(data)));
        assertArrayEquals(data, ((Cbor.Value.Bytes) v).value());

        Cbor.Value t = Cbor.decode(Cbor.encode(Cbor.text("meshu")));
        assertEquals("meshu", ((Cbor.Value.Text) t).value());
    }

    @Test
    void nestedArrayDecodes() {
        // KEYSETS response shape: [0x8A, [[s_id, unit, active, fee, expiry], ...]]
        Cbor.Value inner = Cbor.array(
                Cbor.bytes(new byte[8]), Cbor.uint(0), Cbor.uint(1), Cbor.uint(0), Cbor.uint(0));
        Cbor.Value outer = Cbor.array(Cbor.uint(Op.responseOf(Op.KEYSETS)), Cbor.array(List.of(inner)));
        Cbor.Value decoded = Cbor.decode(Cbor.encode(outer));
        assertInstanceOf(Cbor.Value.Array.class, decoded);
    }
}

File: wallet/src/l3/cbor.test.ts
import { describe, expect, test } from 'vitest';
import {
  array,
  asArray,
  asUint,
  bytes,
  decode,
  encode,
  isBytes,
  text,
  uint,
} from './cbor';
import { Envelope } from './envelope';

const hex = (b: Uint8Array) => Array.from(b).map(x => x.toString(16).padStart(2, '0')).join('');
const unhex = (s: string) => new Uint8Array(s.match(/.{2}/g)!.map(x => parseInt(x, 16)));

describe('CBOR codec (port of core/Cbor.java, TESTVECTORS conformance)', () => {
  // Vector 8: ERROR response must encode to exactly 7 bytes.
  test('vector 8: ERROR [0xFE, 20001, 0] → 8318fe194e2100', () => {
    const env = Envelope.of(0xfe, uint(20001), uint(0));
    expect(hex(env.encode())).toBe('8318fe194e2100');
    expect(env.encode().length).toBe(7);
  });

  test('vector 4: MINT_QUOTE request → exact 42 bytes', () => {
    const pk = new Uint8Array(33);
    pk[0] = 0x02;
    pk.fill(0xa1, 1);
    const env = Envelope.of(0x01, uint(1), uint(1000), uint(0), bytes(pk));
    const expected = '8501011903e800582102' + 'a1'.repeat(32);
    expect(hex(env.encode())).toBe(expected);
    expect(env.encode().length).toBe(42);
  });

  test('canonical minimal int encoding', () => {
    expect(hex(encode(uint(1000)))).toBe('1903e8');
    expect(hex(encode(uint(23)))).toBe('17');
    expect(hex(encode(uint(24)))).toBe('1818');
    expect(hex(encode(uint(0)))).toBe('00');
  });

  test('envelope round trip; opcode is first ELEMENT', () => {
    const env = Envelope.of(0x0d, uint(1), bytes(new Uint8Array(32)), array(uint(0x0d)));
    const back = Envelope.decode(env.encode());
    expect(back.opcode).toBe(0x0d);
    expect(back.fields.length).toBe(3);
  });

  test('unknown trailing array elements ignored (§5.2 forward compat)', () => {
    const v2 = encode(array(uint(0x0a), uint(0), uint(999)));
    const env = Envelope.decode(v2);
    expect(env.opcode).toBe(0x0a);
    expect(env.fields.length).toBe(2);
  });

  test('rejects indefinite-length items (§5.2)', () => {
    expect(() => decode(unhex('9f01ff'))).toThrow(/indefinite/);
  });

  test('rejects maps (§5.1)', () => {
    expect(() => decode(unhex('a10102'))).toThrow(/major type/);
  });

  test('rejects negative ints', () => {
    expect(() => decode(unhex('20'))).toThrow();
  });

  test('rejects trailing bytes', () => {
    expect(() => decode(unhex('0102'))).toThrow(/trailing/);
  });

  // ---- adversarial (review C1 class: attacker-controlled lengths must not allocate)

  test('rejects oversized byte-string length', () => {
    // 0x5A = byte string w/ 4-byte length claiming 0x10000000; input is tiny.
    expect(() => decode(unhex('5a1000000041'))).toThrow(/exceeds/);
  });

  test('rejects oversized text length', () => {
    expect(() => decode(unhex('7a7fffffff6162'))).toThrow(/exceeds/);
  });

  test('rejects huge array count', () => {
    expect(() => decode(unhex('9a7fffffff01'))).toThrow(/exceeds input size/);
  });

  test('rejects truncated length header', () => {
    expect(() => decode(unhex('5a00'))).toThrow(/truncated/);
  });

  test('rejects excessive nesting (review fix: stack exhaustion)', () => {
    // array(1) × 100, terminated by uint 0 — 100 bytes, 100 stack frames.
    const deep = new Uint8Array(100).fill(0x81);
    deep[99] = 0x00;
    expect(() => decode(deep)).toThrow(/nesting/);
  });

  test('legitimate nesting well under the cap decodes fine', () => {
    // 10 levels of array(1) around uint 0.
    let v = uint(0);
    for (let i = 0; i < 10; i++) v = array(v);
    let d = decode(encode(v));
    for (let i = 0; i < 10; i++) d = asArray(d, `level ${i}`)[0]!;
    expect(d).toEqual({ t: 'uint', v: 0n });
  });

  test('round trip: bytes and text', () => {
    const data = new Uint8Array([1, 2, 3]);
    const b = decode(encode(bytes(data)));
    expect(isBytes(b) && Array.from(b.v)).toEqual([1, 2, 3]);
    const t = decode(encode(text('meshu')));
    expect(t).toEqual({ t: 'text', v: 'meshu' });
  });

  test('nested array decode', () => {
    const inner = array(bytes(new Uint8Array(8)), uint(0), uint(1), uint(0), uint(0));
    const outer = array(uint(0x8a), inner);
    const decoded = decode(encode(outer));
    const items = asArray(decoded, 'outer');
    expect(items.length).toBe(2);
    expect(asUint(asArray(items[1]!, 'inner')[2]!, 'active')).toBe(1n);
  });
});
